File: solution/day12/src/main/java/greeting/functional/GreeterFactory.java
package greeting.functional;

import java.util.Map;
import java.util.function.Supplier;

public class GreeterFactory {
    public static final String CASUAL = "casual";
    public static final String INTIMATE = "intimate";
    public static final String FORMAL = "formal";
    private static final Map<String, Greeter> mapping = Map.of(
            CASUAL, () -> "Sup bro?",
            FORMAL, () -> "Good evening, sir.",
            INTIMATE, () -> "Hello Darling!"
    );

    public static Greeter create(String formality) {
        return mapping.getOrDefault(formality, create());
    }

    public static Greeter create() {
        return () -> "Hello.";
    }

    public interface Greeter extends Supplier<String> {
    }
}

File: solution/c#/Day12/Day12/Functional/GreeterFactory.cs
namespace Day12.Functional
{
    public static class GreeterFactory
    {
        public const string Casual = "casual";
        public const string Intimate = "intimate";
        public const string Formal = "formal";

        private static readonly IReadOnlyDictionary<string, Greeter> Mapping =
            new Dictionary<string, Greeter>
            {
                {Casual, () => "Sup bro?"},
                {Formal, () => "Good evening, sir."},
                {Intimate, () => "Hello Darling!"},
            };

        public static Greeter Create(string formality)
            => Mapping.GetValueOrDefault(formality, () => "Hello.");
    }

    public delegate string Greeter();
}